### PR TITLE
Implement TLS recommendations from RFC 9325

### PR DESCRIPTION
**Pull request by julek-wolfssl**

### CODE REVIEW

**Review by dgarske — APPROVED** [review]
Wonderful work! Thank you @julek-wolfssl

**Comment by dgarske** [comment]
Looks like some possibly valid CI tests that you will need to investigate:

```
./wolfssl/internal.h:311:25: error: "WOLFSSL_HARDEN_TLS" is not defined, evaluates to 0 [-Werror=undef]
[545](https://github.com/wolfSSL/wolfssl/actions/runs/4553922724/jobs/8031113351?pr=6242#step:5:546)
  311 | #define WSSL_HARDEN_TLS WOLFSSL_HARDEN_TLS
[546](https://github.com/wolfSSL/wolfssl/actions/runs/4553922724/jobs/8031113351?pr=6242#step:5:547)
      |                         ^~~~~~~~~~~~~~~~~~
[547](https://github.com/wolfSSL/wolfssl/actions/runs/4553922724/jobs/8031113351?pr=6242#step:5:548)
./wolfssl/internal.h:667:44: note: in expansion of macro 'WSSL_HARDEN_TLS'
[548](https://github.com/wolfSSL/wolfssl/actions/runs/4553922724/jobs/8031113351?pr=6242#step:5:549)
  667 |                                            WSSL_HARDEN_TLS > 112)
[549](https://github.com/wolfSSL/wolfssl/actions/runs/4553922724/jobs/8031113351?pr=6242#step:5:550)
      |                                            ^~~~~~~~~~~~~~~
[550](https://github.com/wolfSSL/wolfssl/actions/runs/4553922724/jobs/8031113351?pr=6242#step:5:551)
./wolfssl/internal.h:985:10: error: #error "NULL ciphers not allowed https://www.rfc-editor.org/rfc/rfc9325#section-4.1"
[551](https://github.com/wolfSSL/wolfssl/actions/runs/4553922724/jobs/8031113351?pr=6242#step:5:552)
  985 |         #error "NULL ciphers not allowed https://www.rfc-editor.org/rfc/rfc9325#section-4.1"
[552](https://github.com/wolfSSL/wolfssl/actions/runs/4553922724/jobs/8031113351?pr=6242#step:5:553)
      |          ^~~~~
[553](https://github.com/wolfSSL/wolfssl/actions/runs/4553922724/jobs/8031113351?pr=6242#step:5:554)
./wolfssl/internal.h:994:10: error: #error "At least the server side has to be authenticated"
[554](https://github.com/wolfSSL/wolfssl/actions/runs/4553922724/jobs/8031113351?pr=6242#step:5:555)
  994 |         #error "At least the server side has to be authenticated"
[555](https://github.com/wolfSSL/wolfssl/actions/runs/4553922724/jobs/8031113351?pr=6242#step:5:556)
      |          ^~~~~
[556](https://github.com/wolfSSL/wolfssl/actions/runs/4553922724/jobs/8031113351?pr=6242#step:5:557)
  CC       wolfcrypt/src/src_libwolfssl_la-hpke.lo
[557](https://github.com/wolfSSL/wolfssl/actions/runs/4553922724/jobs/8031113351?pr=6242#step:5:558)
  CC       wolfcrypt/src/src_libwolfssl_la-coding.lo
```

**Review by ejohnstown — APPROVED** [review]

File: src/internal.c [hunk 1]
```diff
@@ -79,6 +79,16 @@
  *     by default.
  *     https://www.rfc-editor.org/rfc/rfc8446#section-5.5
  *     https://www.rfc-editor.org/rfc/rfc9147.html#name-aead-limits
+ * WOLFSSL_HARDEN_TLS
+ *     Implement the recommendations specified in RFC9325. This macro needs to
+ *     be defined to the desired amount of bits of security. The currently
```
Comment (SparkiDev) [inline src/internal.c:84]: amount -> number

File: src/ssl.c [hunk 2]
```diff
@@ -13959,6 +13959,9 @@ int wolfSSL_DTLS_SetCookieSecret(WOLFSSL* ssl,
 
         case ACCEPT_FIRST_REPLY_DONE :
             if ( (ssl->error = SendServerHello(ssl)) != 0) {
+            #ifdef WOLFSSL_CHECK_ALERT_ON_ERR
+                ProcessReplyEx(ssl, 1); /* See if an alert was sent. */
```
Comment (SparkiDev) [inline src/ssl.c:13963]: When was the alert read in?

File: wolfssl/internal.h [hunk 3]
```diff
@@ -692,7 +717,10 @@
         #if !defined(NO_RSA) && defined(HAVE_ECC)
             #define BUILD_TLS_ECDHE_RSA_WITH_CHACHA20_OLD_POLY1305_SHA256
         #endif
-        #if !defined(NO_DH) && !defined(NO_RSA)
+        #if !defined(NO_DH) && !defined(NO_RSA) && !defined(WSSL_HARDEN_TLS)
```
Comment (SparkiDev) [inline wolfssl/internal.h:720]: Don't want to have WSSL_HARDEN_TLS everywhere so can we add a NO_TLS_DH define?

File: wolfssl/internal.h [hunk 4]
```diff
@@ -1123,12 +1183,29 @@ enum {
 
 /* set minimum DH key size allowed */
 #ifndef WOLFSSL_MIN_DHKEY_BITS
-    #ifdef WOLFSSL_MAX_STRENGTH
+    #if defined(WOLFSSL_HARDEN_TLS) && !defined(WOLFSSL_HARDEN_TLS_NO_PKEY_CHECK)
```
Comment (SparkiDev) [inline wolfssl/internal.h:1186]: I didn't think we allowed DH cipher suites.
Do we need to set the minimum DH bits?

File: wolfssl/wolfcrypt/settings.h [hunk 5]
```diff
@@ -2026,6 +2041,16 @@ extern void uITRON4_free(void *p) ;
     #endif
 #endif
 
+#if defined(WOLFSSL_HARDEN_TLS) && ECC_MIN_KEY_SZ < 224 && \
```
Comment (SparkiDev) [inline wolfssl/wolfcrypt/settings.h:2044]: What about when WOLFSSL_HARDEN_TLS is 128?

**Review by SparkiDev — CHANGES_REQUESTED** [review]

File: src/internal.c (hunk 1 above)
Comment (julek-wolfssl) [reply src/internal.c:84]: Changing.

File: wolfssl/internal.h (hunk 4 above)
Comment (julek-wolfssl) [reply wolfssl/internal.h:1186]: We don't when just using `WOLFSSL_HARDEN_TLS` but if `WOLFSSL_HARDEN_TLS_ALLOW_ALL_CIPHERSUITES` would also be defined then DH ciphersuites could be negotiated.

File: wolfssl/wolfcrypt/settings.h (hunk 5 above)
Comment (julek-wolfssl) [reply wolfssl/wolfcrypt/settings.h:2044]: If a user wants to specify a `ECC_MIN_KEY_SZ` when using `WOLFSSL_HARDEN_TLS=128` then we can allow that (the 128 bits is only a SHOULD in the RFC). The 112 bits requirement is a MUST which is why I mandated only the 112 bit requirement.

File: wolfssl/internal.h (hunk 3 above)
Comment (julek-wolfssl) [reply wolfssl/internal.h:720]: Added a `NO_TLS_DH` define for the applicable checks.

File: wolfssl/wolfcrypt/settings.h (hunk 5 above)
Comment (SparkiDev) [reply wolfssl/wolfcrypt/settings.h:2044]: Thanks!

File: wolfssl/internal.h (hunk 4 above)
Comment (SparkiDev) [reply wolfssl/internal.h:1188]: OK

File: src/ssl.c (hunk 2 above)
Comment (SparkiDev) [reply src/ssl.c:13963]: Reads are typically blocking.
If the connection isn't closed it could block indefinitely.

Comment (SparkiDev) [reply src/ssl.c:13963]: Given that you need to define WOLFSSL_CHECK_ALERT_ON_ERR, then we will let the customer figure out how to handle this.

**Review by SparkiDev — APPROVED** [review]

**Review by JacobBarthelmeh — APPROVED** [review]

**Pull request merged**